### ISSUE TITLE
Model checkpointing for sub-epoch frequency

### ISSUE THREAD

**Issue by Anjum48** [PyTorchLightning/pytorch-lightning#1758]
#### What is your question?
In my `Trainer` I have `val_check_interval=0.25`, which is great and I can see that the validation loop is run 4 times per epoch as expected.

Is there a way for `ModelCheckpoint` to use the validation checks to trigger model saving at a sub-epoch frequency? I can see (by setting `verbose=True`) that the check is only done on epoch end

#### What's your environment?

 - OS: Linux
 - Packaging: pip (bleeding edge)
 - Version 0.7.5

**Comment by ekgren** [comment]
Agree, when training on very large datasets it would be very nice to be able to save checkpoints on validation!

**Comment by williamFalcon** [comment]
i thought this was already the case?
agree that should be fixed

**Comment by artidoro** [comment]
This is a bit of a hack but it seems to work:
https://github.com/PyTorchLightning/pytorch-lightning/issues/1809#issuecomment-629881957

**Comment by Anjum48** [comment]
I think @artidoro has struck on the issue. The second condition is always `True` when `on_validation_end` is called within the same epoch:
https://github.com/PyTorchLightning/pytorch-lightning/blob/3459a546672303204a4ae6efcc2613a90f003903/pytorch_lightning/callbacks/model_checkpoint.py#L214-L216

Perhaps the easiest fix is to set `period=-1` by default? Would that break anything else?

**Comment by awaelchli** [comment]
Currently period is an integer value and the behaviour @Anjum48 described is expected: If we check validation multiple times during one epoch but period=1, we expect only one time to save a checkpoint. So clearly **not a bug**. 

In PL currently it only makes sense to checkpoint at epoch intervals, because the training can anyway not be restored mid-epoch currently. So the edge case where we have val_check_interval < 1 is unfortunate. In my opinion this case should not at all save checkpoints mid epoch and only on (val) epoch end because it otherwise just creates issues with restoring training. 

Maybe @jeremyjordan has also comments about this.

**Comment by jeremyjordan** [comment]
> In my opinion this case should not at all save checkpoints mid epoch and only on (val) epoch end because it otherwise just creates issues with restoring training.

I'm confused by this terminology. As I understand it, **epoch end** refers to when we have made a full pass through the training dataset. **Validation end** would refer to when we have made a full pass through the validation set, which depending on the trainer setting, could happen multiple times per training **epoch**. I think I understand what you're saying though that we should only be saving one checkpoint at the end of an epoch?

I'm also confused about mid-epoch checkpoints. The [docs reference](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_io.py#L33):

> You might want to not only load a model but also continue training it. Use this method to restore the trainer state as well. This will continue from the epoch and global step you last left off.

which implies that we can restart training mid-epoch. However, if you look at the [training loop](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L439) it does appear that we always start at the beginning of an epoch and don't respect the global step loaded from a checkpoint.

**Comment by awaelchli** [comment]
I used the terminology wrong. What I meant to say is simply that currenly we save checkpoints on validation_end (for a good reason) but it might be undesireable to do so in case validation_end happens mid-epoch, because restoring from such a checkpoint  leads to incorrectly restored Trainer state (as you pointed out with global_step). So we need to solve this problem first. 
Second is the conflicting setting of the Trainer arg "val_check_interval" and ModelCheckpoint arg "period" that simply do not work togther.

**Comment by teddykoker** [comment]
What is the status on this? We do support sub-epoch checkpointing, but as @awaelchli mentioned, it doesn't restore to the global checkpoint?

**Comment by williamFalcon** [comment]
yeah... i don't think we can actually restore sub epoch state though?

know of a way to do that? we'd have to pull the shuffle state out of the loaders

**Comment by teddykoker** [comment]
I don't it even matters if train loaders are shuffled, I'm not sure if this is important
